Index: b/example/rahunas.conf.in
===================================================================
--- a/example/rahunas.conf.in	2009-04-01 14:57:53.000000000 +0700
+++ b/example/rahunas.conf.in	2009-04-01 15:04:01.000000000 +0700
@@ -5,8 +5,8 @@
 //////////////////////////////////////////////////////////
 
 main = {
-  conf_dir = "@prefix@/etc/rahunas/rahunas.d"
-  log_file = "@prefix@/var/log/rahunas/rahunas.log"
+  conf_dir = "@sysconfdir@/rahunas/rahunas.d"
+  log_file = "@localstatedir@/log/rahunas/rahunas.log"
 
   bandwidth_shape = "yes"  
   bittorrent_download_max = "512"
Index: b/example/rahunas.default.in
===================================================================
--- a/example/rahunas.default.in	2009-04-01 14:57:53.000000000 +0700
+++ b/example/rahunas.default.in	2009-04-01 15:04:01.000000000 +0700
@@ -5,4 +5,4 @@
 RUN_DAEMON=no
 
 # RahuNAS config file
-RAHUNAS_CONFIG=@prefix@/etc/rahunas/rahunas.conf
+RAHUNAS_CONFIG=@sysconfdir@/rahunas/rahunas.conf
Index: b/example/rahunas-vserver.conf.in
===================================================================
--- a/example/rahunas-vserver.conf.in	2009-04-01 15:50:47.000000000 +0700
+++ b/example/rahunas-vserver.conf.in	2009-04-01 15:51:45.000000000 +0700
@@ -10,6 +10,8 @@
   ignore_mac = "no"
   vserver_ip = "172.30.0.1"
   vserver_fqdn = "authen.rahunas.local"
+  vserver_ports_allow = "80,443"
+  vserver_ports_intercept = "80,8080,8088,3128"
   clients = "172.30.0.0/22"
   excluded = ""
   idle_timeout = "600"
